File: CSharpCodeBase/entities/camera/camera.lua
local Class = require 'utils.hump.class'
local Vector3 = require 'utils.hump.vector3'
local UnityExistsEntity = require 'unity.unityentity'

local CameraEntity = Class({__includes=UnityExistsEntity, init = function (self, gameObject)
      Transform.AddComponent(gameObject, "LuaMapper")
      UnityExistsEntity.init(self, gameObject)
      
      self.targetEntity = nil
      
      self.minHeight = 6
      self.minDistance = 12
      self.maxHeight = 30
      self.maxDistance = 40
      self.targetHeight = 1
      
      self.currentZoom = 0.5
      self._zoom = self.currentZoom
      self.zoomSoftLimit = 0.05
      self.lerpZoomSpeed = 5
      self.lerpLimitSpeed = 20
      
      self.angle = 0
      self.currentAngle = 0
      
      self.scrollSpeed = 3
      
      self.luaMapper:SetAlwaysVisible(true)
end})

function CameraEntity:LoadForEntity(target)
  self.targetHeight = target.height / 2
  self.minHeight = target.config:Get('cameraMinHeight') or 6
  self.minDistance = target.config:Get('cameraMinDistance') or 12
  self.maxHeight = target.config:Get('cameraMaxHeight') or 30
  self.maxDistance = target.config:Get('cameraMaxDistance') or 40 
end

function CameraEntity:SetTargetEntity(entity)
  self.targetEntity = entity
end

function CameraEntity:Update()
  local isGamepad = GameController.inputService:IsGamepad()
  local lookValue = GameController.inputService:GetLookValue()
  local scroll = 0
  if isGamepad and GameController.inputService:RightBumperIsPressed() then
    if math.abs(lookValue.z) > 0.5 then
      scroll = lookValue.z
    end
  else
    scroll = Input.GetMouseScrollValue()
  end
  if scroll ~= 0 then
    local sign = scroll > 0 and 1 or -1
    self._zoom = self._zoom - sign * self.scrollSpeed * GameController.deltaTime
    self._zoom = math.clamp(0 - self.zoomSoftLimit, self._zoom, 1 + self.zoomSoftLimit)
  end
  if isGamepad then
    if GameController.inputService:RightBumperIsPressed() then
      local val = lookValue.x
      if math.abs(val) > 0.5 then
        if val < 0 and self.isGamepadRotateBlocked == nil then
          self.currentAngle = self.currentAngle - math.rad(90)
        elseif self.isGamepadRotateBlocked == nil then
          self.currentAngle = self.currentAngle + math.rad(90)
        end
        self.isGamepadRotateBlocked = true
      else
        self.isGamepadRotateBlocked = nil
      end
    end
  else
    if Input.GetKeyDown(KeyCode.Z) then
      self.currentAngle = self.currentAngle - math.rad(90)
    elseif Input.GetKeyDown(KeyCode.X) then
      self.currentAngle = self.currentAngle + math.rad(90)
    end
  end
  
  self.angle = math.lerp(self.angle, self.currentAngle, GameController.deltaTime * self.lerpLimitSpeed)
end

function CameraEntity:FixedUpdate()
end

function CameraEntity:LateUpdate()
  if self.targetEntity ~= nil then  
    -- Align zoom
    if self._zoom < 0 then 
      self._zoom = math.lerp(self._zoom, 0, GameController.deltaTime * self.lerpLimitSpeed)
    elseif self._zoom > 1 then
      self._zoom = math.lerp(self._zoom, 1, GameController.deltaTime * self.lerpLimitSpeed)
    end
    self.currentZoom = math.lerp(self.currentZoom, self._zoom, self.lerpZoomSpeed * GameController.deltaTime)
    
    -- Calculate current parameters
    local heightValue = math.lerp(self.minHeight, self.maxHeight, self.currentZoom)
    local distanceValue = math.lerp(self.minDistance, self.maxDistance, self.currentZoom)

    if self.heightValue == nil then
      self.heightValue = heightValue
      self.distanceValue = distanceValue
    end

    --heightValue = math.lerp(self.heightValue, heightValue, GameController.deltaTime * 50)
    --distanceValue = math.lerp(self.distanceValue, distanceValue, GameController.deltaTime * 50)

    self.heightValue = heightValue
    self.distanceValue = distanceValue

    local offsetVector = Vector3(0, heightValue, -distanceValue)
    offsetVector:RotateAroundY(self.angle)
    
    -- Setup position and rotation
    local position = self.targetEntity:GetPosition() + offsetVector
    self:SetPosition(position)
    
    -- Setup rotation
    offsetVector.y = offsetVector.y - self.targetHeight
    local rotation = RotationUtils.LookRotation(offsetVector * -1)
    self:SetRotation(rotation)
  end
end

return CameraEntity

File: CSharpCodeBase/entities/camera/camera.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using UnityEngine;
using System.Collections;
using MainGame.core;
//using UnityEngine.EventSystems;
 
 namespace MainGame
 {
    public class Camera : UnityEntity
    {
        protected GameObject targetEntity = null;
        protected float minHeight = 6;
        protected float minDistance = 12;
        protected float maxHeight = 30;
        protected float maxDistance = 40;
        protected float targetHeight = 1;
        protected float currentZoom = 0.5f;
        protected float zoomSoftLimit = 0.05f;
        protected float lerpZoomSpeed = 5;
        protected float lerpLimitSpeed = 20;
        protected float angle = 0;
        protected float currentAngle = 0;
        protected float scrollSpeed = 3;
        protected float heightValue;
        protected float distanceValue;

        private float zoom;

        public void Awake()
        {            
            zoom = currentZoom;
        }         
         
        public void LoadForEntity(GameObject target)
        {
            targetHeight = target.GetComponent<UnityEntity>().height / 2;
            //minHeight = target.GetComponent<Config>().Get("cameraMinHeight")  ||  6;
            //minDistance = target.GetComponent<Config>().Get("cameraMinDistance")  ||  12;
            //maxHeight = target.GetComponent<Config>().Get("cameraMaxHeight")  ||  30;
            //maxDistance = target.GetComponent<Config>().Get("cameraMaxDistance")  ||  40 ;
        }

        public void SetTargetEntity(GameObject entity)
        {
            targetEntity = entity;
        }

        public void Update()
        {
            //var isGamepad = GameController.inputService:IsGamepad();
            //var lookValue = GameController.inputService:GetLookValue();
            int scroll = 0;
            //if(isGamepad  &&  GameController.inputService:RightBumperIsPressed())
            //{ 
                //if(math.abs(lookValue.z) > 0.5)
                //{ 
                    //scroll = lookValue.z;
                //}
            //}else
            //{
                //scroll = Input.GetMouseScrollValue();
            //}
            if(scroll != 0)
            { 
                int sign = scroll > 0  ? 1  : -1;
                zoom = zoom - sign * scrollSpeed * Time.deltaTime;
                zoom = Mathf.Clamp(0 - zoomSoftLimit, zoom, 1 + zoomSoftLimit);
            }
            //if(isGamepad)
            //{ 
                //if(GameController.inputService:RightBumperIsPressed())
                //{ 
                    //var val = lookValue.x;
                    //if(math.abs(val) > 0.5)
                    //{ 
                        //if(val < 0  &&  this.isGamepadRotateBlocked == null)
                        //{ 
                            //this.currentAngle = this.currentAngle - math.rad(90);
                        //}else
                        //{
                            //if this.isGamepadRotateBlocked == null)
                            //{ 
                                //this.currentAngle = this.currentAngle + math.rad(90);
                            //}
                            //this.isGamepadRotateBlocked = true;
                        //}
                    //}
                    //else{
                        //this.isGamepadRotateBlocked = null;
                    //}
                //}
            //}
            //else
            //{
                //if(Input.GetKeyDown(KeyCode.Z))
                //{ 
                    //this.currentAngle = this.currentAngle - math.rad(90);
                //}
                //else
                //{
                    //if Input.GetKeyDown(KeyCode.X))
                    //{ 
                        //this.currentAngle = this.currentAngle + math.rad(90);
                    //}
                //}
             //}
   
            angle = Mathf.Lerp(angle, currentAngle, Time.deltaTime * lerpLimitSpeed);
        }

        public void FixedUpdate()
        {
        }

        public void LateUpdate()
        {
             if (targetEntity != null)
             {   
                //Align zoom
                if (zoom < 0)
                {  
                    zoom = Mathf.Lerp(zoom, 0, Time.deltaTime * lerpLimitSpeed);
                }
                else
                {
                    if (zoom > 1)
                    { 
                        zoom = Mathf.Lerp(zoom, 1, Time.deltaTime * lerpLimitSpeed);
                    }
                    currentZoom = Mathf.Lerp(currentZoom, zoom, lerpZoomSpeed * Time.deltaTime);
     
                    //Calculate current parameters
                    float heightValue = Mathf.Lerp(minHeight, maxHeight, currentZoom);
                    float distanceValue = Mathf.Lerp(minDistance, maxDistance, currentZoom);
                    if (heightValue == null)
                    {
                        this.heightValue = heightValue;
                        this.distanceValue = distanceValue;
                    }
                    //heightValue = math.lerp(this.heightValue, heightValue, GameController.deltaTime * 50)
                    //distanceValue = math.lerp(this.distanceValue, distanceValue, GameController.deltaTime * 50)
                    this.heightValue = heightValue;
                    this.distanceValue = distanceValue;
                    Vector3 offsetVector = new Vector3(0, heightValue, -distanceValue);
                    //offsetVector:RotateAroundY(this.angle);
     
                    //Setup position  &&  rotation
                    Vector3 position = targetEntity.transform.position + offsetVector;
                    SetPosition(position);
     
                    //Setup rotation
                    offsetVector.y = offsetVector.y - targetHeight;
                    Vector3 rotation = /*RotationUtils.LookRotation(offsetVector * -1)*/;
                    self:SetRotation(rotation);
                }
             }
         }
    }
 }